### PR TITLE
Bbonev/filter basis functions

### PR DESCRIPTION
**Pull request by bonevbs**
Refactoring of code to support flexible choise of filter basis functions for spherical convolutions.

### CODE REVIEW

File: torch_harmonics/convolution.py [hunk 1]
```diff
@@ -257,7 +172,7 @@ def _precompute_convolution_tensor_s2(
         phi = torch.arctan2(y, x) + torch.pi
 
         # find the indices where the rotated position falls into the support of the kernel
-        iidx, vals = kernel_handle(theta, phi)
+        iidx, vals = filter_basis.compute_support_vals(theta, phi, r_cutoff=theta_cutoff)
```
Comment (azrael417) [inline torch_harmonics/convolution.py:175]: The naming is weird I think, since radial and angular cutoff are technically different. Can we streamline that somehow?

File: torch_harmonics/convolution.py [hunk 2]
```diff
@@ -275,7 +190,7 @@ def _precompute_convolution_tensor_s2(
     else:
         quad_weights = 2.0 * torch.pi * torch.from_numpy(win).float().reshape(-1, 1) / nlon_in
     out_vals = _normalize_convolution_tensor_s2(
-        out_idx, out_vals, in_shape, out_shape, kernel_shape, quad_weights, transpose_normalization=transpose_normalization, merge_quadrature=merge_quadrature
+        out_idx, out_vals, in_shape, out_shape, kernel_size, quad_weights, transpose_normalization=transpose_normalization, merge_quadrature=merge_quadrature
```
Comment (azrael417) [inline torch_harmonics/convolution.py:193]: should we name it kernel_sizes, to make it more clear that you can also do anisotropic kernels?

File: torch_harmonics/filter_basis.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+def get_filter_basis(kernel_shape: Union[int, List[int], Tuple[int, int]], basis_type: str):
+    """Factory function to generate the appropriate filter basis"""
+
+    if basis_type == "piecewise linear":
+        return PiecewiseLinearFilterBasis(kernel_shape=kernel_shape)
+    else:
+        raise ValueError(f"Unknown basis_type {basis_type}")
+
+
+class AbstractFilterBasis(metaclass=abc.ABCMeta):
```
Comment (azrael417) [inline torch_harmonics/filter_basis.py:48]: why not just call it FilterBasis ?

**Review by azrael417 — APPROVED** [review]
LGTM, just naming nitpicking.

File: tests/test_convolution.py [hunk 4]
```diff
@@ -217,30 +167,32 @@ def setUp(self):
...
+            [8, 4, 2, (16, 32), (16, 32), [3], "piecewise linear", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (8, 16), (16, 32), [5], "piecewise linear", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (12, 24), (24, 48), [3, 3], "piecewise linear", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (12, 24), (24, 48), [4, 3], "piecewise linear", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (12, 24), (24, 48), [2, 2], "morlet", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (8, 8), (16, 24), [3], "piecewise linear", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (6, 12), (18, 36), [7], "piecewise linear", "mean", "equiangular", "equiangular", True, 1e-4],
+            [8, 4, 2, (8, 16), (16, 32), [5], "piecewise linear", "mean", "equiangular", "legendre-gauss", True, 1e-4],
+            [8, 4, 2, (8, 16), (16, 32), [5], "piecewise linear", "mean", "legendre-gauss", "equiangular", True, 1e-4],
+            [8, 4, 2, (8, 16), (16, 32), [5], "piecewise linear", "mean", "legendre-gauss", "legendre-gauss", True, 1e-4],
```
Comment (azrael417) [inline tests/test_convolution.py:195]: we should also test individual and none normalization, right?

File: torch_harmonics/examples/models/_layers.py [hunk 5]
```diff
@@ -258,123 +250,19 @@ def __init__(self,
...
-        with torch.autocast(device_type="cuda", enabled=False):
-            x = self.inverse_transform(x)
-
-        if hasattr(self, "bias"):
-            x = x + self.bias
-        x = x.type(dtype)
-
-        return x, residual
-
-class FactorizedSpectralConvS2(nn.Module):
```
Comment (azrael417) [inline torch_harmonics/examples/models/_layers.py:308]: is this completely gone?

**Review by azrael417 — APPROVED** [review]

File: torch_harmonics/examples/models/_layers.py (hunk 5 above)
Comment (bonevbs) [reply torch_harmonics/examples/models/_layers.py:308]: yes - this is to simplify example architectures and remove dependencies

File: torch_harmonics/filter_basis.py (hunk 3 above)
Comment (bonevbs) [reply torch_harmonics/filter_basis.py]: this has been addressed

File: tests/test_convolution.py (hunk 4 above)
Comment (bonevbs) [reply tests/test_convolution.py:202]: thanks - i added those
